torch/csrc/api/src/optim/adam.cpp

```diff
@@ -12,57 +12,171 @@
+      auto exp_avg = state.exp_avg();
+      // Exponential moving average of squared gradient values
+      auto exp_avg_sq = state.exp_avg_sq();
+      Tensor max_exp_avg_sq;
+      if(options.amsgrad()) {
+        // Maintains max of all exp. moving avg. of sq. grad. values
+        max_exp_avg_sq = state.max_exp_avg_sq();
+      }
+      state.step(state.step()+1);
+      auto bias_correction1 = 1 - std::pow(options.beta1(), state.step());
+      auto bias_correction2 = 1 - std::pow(options.beta2(), state.step());
 
-    if (options.weight_decay() > 0) {
-      NoGradGuard guard;
-      p.grad() = p.grad() + options.weight_decay() * p;
-    }
+      if(options.weight_decay() != 0) {
+        grad = grad.add(p.data(), options.weight_decay());
+      }
 
-    auto& exp_average = buffer_at(exp_average_buffers, i);
-    auto& exp_average_sq = buffer_at(exp_average_sq_buffers, i);
-
-    buffer_at(step_buffers, i) += 1;
-    const auto bias_correction1 =
-        1 - std::pow(options.beta1(), buffer_at(step_buffers, i));
-    const auto bias_correction2 =
-        1 - std::pow(options.beta2(), buffer_at(step_buffers, i));
-
-    exp_average.mul_(options.beta1()).add_(p.grad(), 1 - options.beta1());
-    exp_average_sq.mul_(options.beta2())
-        .addcmul_(p.grad(), p.grad(), 1 - options.beta2());
-
-    Tensor denom;
-    if (options.amsgrad()) {
-      auto& max_exp_average_sq = buffer_at(max_exp_average_sq_buffers, i);
-      max_exp_average_sq = torch::max(max_exp_average_sq, exp_average_sq);
-      denom = max_exp_average_sq / bias_correction2;
-    } else {
-      denom = exp_average_sq / bias_correction2;
-    }
+      // Decay the first and second moment running average coefficient
+      exp_avg.mul_(options.beta1()).add_(grad, 1 - options.beta1());
+      exp_avg_sq.mul_(options.beta2()).addcmul_(grad, grad, 1 - options.beta2());
 
-    const auto step_size =
-        options.learning_rate() / bias_correction1;
+      Tensor denom;
+      if(options.amsgrad()) {
+        // Maintains the maximum of all 2nd moment running avg. till now
+        at::max_out(max_exp_avg_sq, exp_avg_sq, max_exp_avg_sq);
+        // Use the max. for normalizing running avg. of gradient
+        denom = (max_exp_avg_sq.sqrt() / sqrt(bias_correction2)).add_(options.eps());
+      } else {
+        denom = (exp_avg_sq.sqrt() / sqrt(bias_correction2)).add_(options.eps());
+      }
 
-    NoGradGuard guard;
-    p.addcdiv_(exp_average, denom.sqrt() + options.eps(), -step_size);
+      auto step_size = options.lr() / bias_correction1;
+      p.data().addcdiv_(exp_avg, denom, -step_size);
+    }
   }
 }
 
+void Adam::add_parameters(const std::vector<Tensor>& parameters) {
+  param_groups_.emplace_back(OptimizerParamGroup(parameters, defaults_->clone()));
+}
+
+const std::vector<Tensor>& Adam::parameters() const noexcept {
+  return param_groups_.at(0).params();
+}
+
+std::vector<Tensor>& Adam::parameters() noexcept {
+  return param_groups_.at(0).params();
+}
```

Comment:
The content of the above three functions `add_parameters`  / const ref `parameters` / non-const ref `parameters` seem to always be the same for all optimizers, and maybe we should create functions similar to `_size_new_design` and have all optimizers call the common functions instead :D